Make resize_as_ generic, so XLA works.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26719 Tests for fallback boxed dispatch.
* #26368 Add boxed fallback
* #26367 Refactor dispatch structure so fallback code lives inline.
* #26857 Change calling convention of ATenDispatch from getOp to callUnboxed.
* #26806 Add some missing constructors to IValue.
* **#26809 Make resize_as_ generic, so XLA works.**

resize_as_ shouldn't do multiple dispatch on its second argument.  Because it
currently has per CPU/CUDA dispatch, however, it will do proper dispatch on all
arguments. Bad!

There is only a very minor downside to this patch which is we have an extra
dynamic dispatch now.

Thank you Ailing for reporting this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D17581324](https://our.internmc.facebook.com/intern/diff/D17581324)

ezyang committed:
Make resize_as_ generic, so XLA works.

resize_as_ shouldn't do multiple dispatch on its second argument.  Because it
currently has per CPU/CUDA dispatch, however, it will do proper dispatch on all
arguments. Bad!

There is only a very minor downside to this patch which is we have an extra
dynamic dispatch now.

Thank you Ailing for reporting this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Make resize_as_ generic, so XLA works."


resize_as_ shouldn't do multiple dispatch on its second argument.  Because it
currently has per CPU/CUDA dispatch, however, it will do proper dispatch on all
arguments. Bad!

There is only a very minor downside to this patch which is we have an extra
dynamic dispatch now.

Thank you Ailing for reporting this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ailzhang commented on aten/src/ATen/test/extension_backend_test.cpp:
```diff
@@ -62,3 +66,42 @@ TEST(BackendExtensionTest, TestRegisterOp) {
+
+// This is a "shitty" version of lazy tensor, where we run every operator by
+// running it on JIT
+struct ViaJitTensorImpl : public c10::TensorImpl {
```
Hmmm I'm not sure whether it's necessary to write a hacky lazy tensor here. 
But I usually add my test in https://github.com/pytorch/pytorch/blob/master/test/cpp_extensions/msnpu_extension.cpp#L29. That's exactly how XLA is registered these days.

ezyang replied:
This is my bad, the testing changes shouldn't be in this diff

ailzhang reviewed (approved):
Thanks Ed! 
Only a minor comment about testing. ;)

ezyang committed:
Update on "Make resize_as_ generic, so XLA works."


resize_as_ shouldn't do multiple dispatch on its second argument.  Because it
currently has per CPU/CUDA dispatch, however, it will do proper dispatch on all
arguments. Bad!

There is only a very minor downside to this patch which is we have an extra
dynamic dispatch now.

Thank you Ailing for reporting this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -16,8 +16,14 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
 
-Tensor& resize_as_cpu_(Tensor& self, const Tensor& the_template) {
-  Tensor& result = resize_cpu_(self, the_template.sizes());
+// I am a naughty man
```
Please clarify? It's just that we shouldn't be resizing sparse tensors, right?

ezyang replied:
Oh, it's because I stuck the forward declaration here, instead of in a header

ezyang replied:
It's soooo naughty that ROCm hates it. I guess I'll need to find another way to do this.

ezyang committed:
Update on "Make resize_as_ generic, so XLA works."


resize_as_ shouldn't do multiple dispatch on its second argument.  Because it
currently has per CPU/CUDA dispatch, however, it will do proper dispatch on all
arguments. Bad!

There is only a very minor downside to this patch which is we have an extra
dynamic dispatch now.

Thank you Ailing for reporting this problem.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
